src/rgw/rgw_sal_config.h

```diff
@@ -0,0 +1,300 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab ft=cpp
+
+/*
+ * Ceph - scalable distributed file system
+ *
+ * Copyright (C) 2022 Red Hat, Inc.
+ *
+ * This is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License version 2.1, as published by the Free Software
+ * Foundation. See file COPYING.
+ *
+ */
+
+#pragma once
+
+#include <span>
+#include <string>
+#include "rgw_sal_fwd.h"
+
+class DoutPrefixProvider;
+class optional_yield;
+struct RGWPeriod;
+struct RGWPeriodConfig;
+struct RGWRealm;
+struct RGWZoneGroup;
+struct RGWZoneParams;
+
+namespace rgw::sal {
+
+/// Results of a listing operation
+template <typename T>
+struct ListResult {
+  /// The subspan of the input entries that contain results
+  std::span<T> entries;
+  /// The next marker to resume listing, or empty
+  std::string next;
+};
+
+class RealmWriter;
+class ZoneGroupWriter;
+class ZoneWriter;
+
+/// Storage abstraction for realm/zonegroup/zone configuration
+class ConfigStore {
+ public:
+  virtual ~ConfigStore() {}
+
+  /// @group Realm
+  ///@{
+
+  /// Set the cluster-wide default realm id
+  virtual int write_default_realm_id(const DoutPrefixProvider* dpp,
+                                     optional_yield y, bool exclusive,
+                                     std::string_view realm_id) = 0;
+  /// Read the cluster's default realm id
+  virtual int read_default_realm_id(const DoutPrefixProvider* dpp,
+                                    optional_yield y,
+                                    std::string& realm_id) = 0;
+  /// Delete the cluster's default realm id
+  virtual int delete_default_realm_id(const DoutPrefixProvider* dpp,
+                                      optional_yield y) = 0;
+
+  /// Create a realm
+  virtual int create_realm(const DoutPrefixProvider* dpp,
+                           optional_yield y, bool exclusive,
+                           const RGWRealm& info,
+                           std::unique_ptr<RealmWriter>* writer) = 0;
+  /// Read a realm by id
+  virtual int read_realm_by_id(const DoutPrefixProvider* dpp,
+                               optional_yield y,
+                               std::string_view realm_id,
+                               RGWRealm& info,
+                               std::unique_ptr<RealmWriter>* writer) = 0;
```

Comment:
Why do the read* functions need to take a writer?  The writers don't seem to have any read methods.

Reply:
in rados, we use `RGWObjVersionTracker` to implement 'atomic' read-modify-writes. by passing the `objv` to reads, we read the cls_version number we need for the successive writes

this `writer` param is a pointer because it's optional - if the caller doesn't need to do a read-modify-write, they can pass nullptr

it would help if you could see the `RadosConfigStore` - i have that implemented, but wasn't able to push it yesterday before my power went out.. so hopefully it's not lost